src/nvim/fold.c

```diff
@@ -1336,7 +1336,7 @@ void foldMarkAdjust(win_T *wp, linenr_T line1, linenr_T line2, linenr_T amount,
   // If deleting marks from line1 to line2, but not deleting all those
   // lines, set line2 so that only deleted lines have their folds removed.
   if (amount == MAXLNUM && line2 >= line1 && line2 - line1 >= -amount_after) {
-    line2 = line1 - amount_after - 1;
```

Comment:
This matches the comment above: "only deleted lines have their folds removed". If you change this to `line2 += amount_after` this now becomes "only non-deleted lines have their folds removed".

Reply:
As per `mark_adjust()`, if `amount = MAXLNUM` then this should remove all marks between `line1` and `line2`.

With the provided test case, `line1 = 1`, `line2 = 4`, `amount = MAXLNUM`, `amount_after = 1`. This results in this line modifying `line2 = -1` which I believe then causes the issue.

I'm not 100% on my solution, but adding `amount_after` results in `line = 5` which is the range in which we need to remove all marks.

Reply:
If `amount_after` is positive (indicating we have added lines), then this line always executes.

Reply:
Here's an alternative fix that doesn't change `fold.c`:

```diff
diff --git a/src/nvim/api/buffer.c b/src/nvim/api/buffer.c
index d3895d31c..3e6ca5e67 100644
--- a/src/nvim/api/buffer.c
+++ b/src/nvim/api/buffer.c
@@ -488,11 +488,11 @@ void nvim_buf_set_lines(uint64_t channel_id, Buffer buffer, Integer start, Integ
   // changed range, and move any in the remainder of the buffer.
   // Only adjust marks if we managed to switch to a window that holds
   // the buffer, otherwise line numbers will be invalid.
-  mark_adjust((linenr_T)start,
-              (linenr_T)(end - 1),
-              MAXLNUM,
-              (linenr_T)extra,
-              kExtmarkNOOP);
+  // Call this as:
+  //    - deleting lines start to end
+  //    - adding lines start to end + extra
+  mark_adjust((linenr_T)start, (linenr_T)end, MAXLNUM, (linenr_T)(to_replace), kExtmarkNOOP);
+  mark_adjust((linenr_T)start, (linenr_T)(start - 1), MAXLNUM, (linenr_T)to_replace + (linenr_T)extra, kExtmarkNOOP);
 
   extmark_splice(curbuf, (int)start - 1, 0, (int)(end - start), 0,
                  deleted_bytes, (int)new_len, 0, inserted_bytes,

```

Reply:
I also thought about using `mark_adjust_nofold()` and two calls to `foldMarkAdjust` here.